Update toml dependency to use tomli instead of toml

Taaku18 opened the pull request:
The toml library beanie uses is years out of date, and [doesn't support the newer toml syntax (from 5 years ago...) of mixing types in arrays](https://github.com/uiri/toml/issues/270). 

When running migrations, beanie uses the toml library to parse config from pyproject.toml. If the pyproject file contains mixed types, such as this valid tox config:

```
[tool.tox.env.precommit]
description = "run pre-commit"
skip_install = true
deps = ["pre-commit"]
commands = [["pre-commit", { replace = "posargs", default = ["run", "--all-files"], extend = true} ]]
```

`beanie migrate -p [...]` will error out with:

    File ".venv/lib/python3.12/site-packages/toml/decoder.py", line 1002, in load_array
      a[b] = a[b] + ',' + a[b + 1]
                          ~^^^^^^^
    IndexError: list index out of range

This PR changes the dependency from toml to tomli (and tomli-w for GH script). It's the most used toml library that provides backports of the builtin tomllib to py3.8.

staticxterm reviewed (changes requested):
Hi, thank you for the PR. Two suggestions:

- specify and pin down the dependency versions
- don't use tomli for Python 3.11 and greater

Taaku18 commented:
> Hi, thank you for the PR. Two suggestions:
> 
>     * specify and pin down the dependency versions
> 
>     * don't use tomli for Python 3.11 and greater

I've updated the dependency list to include the version locks as requested.
Also switched to importlib for >=3.11; used `sys.version_info` check instead of your suggested `try...except ImportError...` to satisfy mypy.

staticxterm reviewed (approved):
Looks good. Thank you.

Riverfount reviewed (approved):
Great job.
